src/PwrDrvr.LambdaDispatch.Router/LambdaInstanceManager.cs

```diff
@@ -232,60 +218,118 @@ private async Task ManageCapacity()
         var message = await _capacityChannel.Reader.ReadAsync(cts.Token);
         var pendingRequests = message.PendingRequests;
         var runningRequests = message.RunningRequests;
-
-        // Mark that we read a message
-        noMessagesRead = false;
-
-        trailingAverage.Add(ComputeDesiredInstanceCount(pendingRequests, runningRequests));
+        var requestsPerSecondEWMA = message.RequestsPerSecondEWMA;
+        var requestDurationEWMA = message.RequestDurationEWMA;
 
         // Rip through any other messages, up to a limit of 100
         // But keep reading if we keep ending on a message with no pending or running requests
         var messagesToRead = 100;
+        var gotNonZeroPendingOrRunning = pendingRequests != 0 || runningRequests != 0;
         while (_capacityChannel.Reader.TryRead(out var nextMessage)
           && (--messagesToRead > 0 || nextMessage.PendingRequests == 0 && nextMessage.RunningRequests == 0))
         {
           pendingRequests = nextMessage.PendingRequests;
           runningRequests = nextMessage.RunningRequests;
+          requestsPerSecondEWMA = nextMessage.RequestsPerSecondEWMA;
+          requestDurationEWMA = nextMessage.RequestDurationEWMA;
+
+          if (pendingRequests != 0 || runningRequests != 0)
+          {
+            gotNonZeroPendingOrRunning = true;
+          }
+        }
+
+        // Start with the simple count
+        var simpleScalerDesiredInstanceCount = _capacityManager.SimpleDesiredInstanceCount(pendingRequests, runningRequests);
+        var newDesiredInstanceCount = simpleScalerDesiredInstanceCount;
 
-          trailingAverage.Add(ComputeDesiredInstanceCount(pendingRequests, runningRequests));
+        // Override the simple scale-from zero logic (which uses running and pending requests only)
+        // with the EWMA data, if available
+        // If we got a
+        int? ewmaScalerDesiredInstanceCount = null;
+        if (requestsPerSecondEWMA > 0 && requestDurationEWMA > 0)
+        {
+          ewmaScalerDesiredInstanceCount = _capacityManager.EwmaDesiredInstanceCount(requestsPerSecondEWMA, requestDurationEWMA);
         }
 
-        var averageCount = trailingAverage.Average;
-        // If the entire average count is 0 then we want 0 instances
-        // OR: If there we read all the messages and there are no pending or running requests
-        //     in the last read message then we want 0 instances - this helps avoid
-        //     5 second waits to stop instances when load disappears
-        var newDesiredInstanceCount = averageCount == 0
-          || (messagesToRead != 0 && pendingRequests == 0 && runningRequests == 0)
-            ? 0 : (int)Math.Ceiling(averageCount);
```

Comment:
THIS WAS THE ISSUE (2)!

- This is where the scale to zero would happen too frequently
- It happened because it didn't check if there were any `pendingRequests > 0` or `runningRequests > 0` - but even that was not sufficient to fix it as there are many cases where zeros are written into the channel when under heavy load
- There needs to be a "time since last non-zero" check that skips this scale to zero block if there are incoming messages in, say, the last `100 ms`